### PR TITLE
Add custom formatters for messages

### LINKED ISSUE #24: More readable logs
screen shot 2018-05-08 at 16 31 17

screen shot 2018-05-08 at 16 30 32

It would make it much more readable for me if the default logs are the second picture.

### PR DESCRIPTION
**Pull request by manoskouvarakis**
fixes #24

### CODE REVIEW

File: lib/options.js [hunk 1]
```diff
@@ -4,9 +4,11 @@ const out = require('./appenders/console');
 
 module.exports = (options = {}) => {
   const color = typeof options.color === 'boolean' ? options.color : true;
-  const defaultLogger = out({ color });
+  const styles = typeof options.styles === 'object' ? options.styles : ['extended'];
```
Comment (nikostoulas) [inline lib/options.js]: What are you trying here? Maybe 
```js
const {isEmpty} =require('lodash');
!isEmpty(options.styles)
```
 ?

File: types/riviere.d.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 export function riviere(options?: {
   color?: boolean,
+  styles: Array<string>,
```
Comment (nikostoulas) [inline types/riviere.d.ts]: Can you also update the README with what is the defaullt and what is this?

File: test/lib/formatters/FormatterFactoryTest.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
+const formatterFactory = require('../../../lib/formatters/FormatterFactory');
+const SimpleFormatter = require('../../../lib/formatters/SimpleFormatter');
+const ExtendedFormatter = require('../../../lib/formatters/ExtendedFormatter');
+const JsonFormatter = require('../../../lib/formatters/JsonFormatter');
+
+const should = require('should');
+
+describe('Test Formatter Factory', function() {
```
Comment (SokratisVidros) [inline test/lib/formatters/FormatterFactoryTest.js]: Nit: Prefer arrow functions

File: lib/formatters/FormatterFactory.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      return new SimpleFormatter(color, appendDate, requestType);
+    }
+    case 'extended': {
+      return new ExtendedFormatter(color, appendDate, requestType);
+    }
+    case 'json': {
+      return new JsonFormatter(color, appendDate, requestType);
+    }
+    default:
+      return null;
```
Comment (SokratisVidros) [inline lib/formatters/FormatterFactory.js]: Nit: Maybe throw an error to early catch problematic setup?

File: lib/formatters/JsonFormatter.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+const BaseFormatter = require('./BaseFormatter');
+
+class JsonFormatter extends BaseFormatter {
+  formatObject(obj) {
+    const defaultObject = {
+      severity: 'info',
+      message: 'Test message',
```
Comment (SokratisVidros) [inline lib/formatters/JsonFormatter.js]: Nit: Please remove!

Comment (manoskouvarakis) [reply lib/formatters/JsonFormatter.js]: Done

File: lib/formatters/FormatterFactory.js (hunk 4 above)
Comment (manoskouvarakis) [reply lib/formatters/FormatterFactory.js]: Added "MissingFormatter" to inform for the wrong parameter without breaking other formatters

File: test/lib/formatters/FormatterFactoryTest.js (hunk 3 above)
Comment (manoskouvarakis) [reply test/lib/formatters/FormatterFactoryTest.js]: Done

File: types/riviere.d.ts (hunk 2 above)
Comment (manoskouvarakis) [reply types/riviere.d.ts]: Done

File: lib/options.js (hunk 1 above)
Comment (manoskouvarakis) [reply lib/options.js]: Fixed

File: README.md [hunk 6]
```diff
@@ -255,6 +256,23 @@ Log colored log messages. Defaults to: `true`
 }
 ```
 
+<a name="options_styles"></a>
```
Comment (nikostoulas) [inline README.md:259]: 👏

File: examples/hello.js [hunk 7]
```diff
@@ -3,7 +3,7 @@ const riviere = require('../index');
 
 const app = new Koa();
 
-app.use(riviere());
+app.use(riviere({ styles: ['extended-11', 'json'] }));
```
Comment (nikostoulas) [inline examples/hello.js]:
```suggestion
app.use(riviere({ styles: ['extended', 'json'] })); 
``` ??

**Review by nikostoulas — APPROVED** [review]

File: examples/hello.js [hunk 8]
```diff
@@ -3,7 +3,8 @@ const riviere = require('../index');
 
 const app = new Koa();
 
-app.use(riviere());
+app.use(riviere({ styles: ['extended', 'json'] })); 
+``` ??
```
Comment (alkiskal) [inline examples/hello.js]: typo?

File: lib/adapters/default.js [hunk 9]
```diff
@@ -1,12 +1,11 @@
 const url = require('url');
 const uuidv4 = require('uuid/v4');
 
-const transformersPath = '../transformers';
-const mapOutRes = require(`${transformersPath}/mapOutRes`);
-const mapInReq = require(`${transformersPath}/mapInReq`);
-const mapOutReq = require(`${transformersPath}/mapOutReq`);
-const mapInRes = require(`${transformersPath}/mapInRes`);
-const mapError = require(`${transformersPath}/mapError`);
+const mapOutRes = require('../transformers/mapOutRes');
```
Comment (alkiskal) [inline lib/adapters/default.js]: Nit: maybe make an index file and require them in one go
`const { mapOutRes, mapInReq etc } = require('../transformers')`

File: lib/formatters/BaseFormatter.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  getInRequestPrefix(obj) {
+    let prefix = `[${obj.requestId}] ${inReq.MSG} ${obj.method} ${obj.path}`;
+    if (this.appendDate) {
+      prefix = `[${new Date().toISOString()}] ${prefix}`;
+    }
+    if (this.color) {
+      prefix = chalk[[colors[inReq.TAG]]](prefix);
+    }
+
+    return `${prefix}`;
```
Comment (alkiskal) [inline lib/formatters/BaseFormatter.js]: Is prefix something else than string here?

File: lib/formatters/BaseFormatter.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+  getOutRequestPrefix(obj) {
+    let prefix = `[${obj.requestId}] ${outReq.MSG} ${obj.method} ${obj.href}`;
+    if (this.appendDate) {
+      prefix = `[${new Date().toISOString()}] ${prefix}`;
+    }
+    if (this.color) {
+      prefix = chalk[colors[outReq.TAG]](prefix);
+    }
+    return `${prefix}`;
```
Comment (alkiskal) [inline lib/formatters/BaseFormatter.js]: same

File: lib/formatters/BaseFormatter.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    }
+    if (this.color) {
+      prefix = obj.status < 400 ? chalk[colors[outRes.TAG]](prefix) : chalk[colors.err](prefix);
+    }
+
+    return `${prefix}`;
+  }
+
+  getInRequestHealthPrefix(obj) {
+    let msg = `${inReqH.MSG} ${obj.method} ${obj.path}`;
```
Comment (alkiskal) [inline lib/formatters/BaseFormatter.js]: Nit: these methods are almost the same, please use the same name variables msg -> prefix, if it doesn't have any special meaning

File: examples/hello.js (hunk 8 above)
Comment (manoskouvarakis) [reply examples/hello.js]: Yes. Fixed

File: lib/formatters/BaseFormatter.js (hunk 12 above)
Comment (manoskouvarakis) [reply lib/formatters/BaseFormatter.js]: Correct. Thanks! Fixed

File: lib/formatters/BaseFormatter.js (hunk 11 above)
Comment (manoskouvarakis) [reply lib/formatters/BaseFormatter.js]: Fixed

File: lib/formatters/BaseFormatter.js (hunk 10 above)
Comment (manoskouvarakis) [reply lib/formatters/BaseFormatter.js]: Fixed

File: lib/adapters/default.js (hunk 9 above)
Comment (manoskouvarakis) [reply lib/adapters/default.js]: Done

File: lib/adapters/default.js [hunk 13]
```diff
@@ -1,12 +1,7 @@
 const url = require('url');
 const uuidv4 = require('uuid/v4');
 
-const transformersPath = '../transformers';
-const mapOutRes = require(`${transformersPath}/mapOutRes`);
-const mapInReq = require(`${transformersPath}/mapInReq`);
-const mapOutReq = require(`${transformersPath}/mapOutReq`);
-const mapInRes = require(`${transformersPath}/mapInRes`);
-const mapError = require(`${transformersPath}/mapError`);
+const { mapError, mapInReq, mapInRes, mapOutReq, mapOutRes } = require('../transformers/transformers.js');
```
Comment (alkiskal) [inline lib/adapters/default.js]: no need to add the extension

File: lib/appenders/console.js [hunk 14]
```diff
@@ -1,25 +1,39 @@
...
-        process.stdout.write(`${fmtErr}\n`);
+    if (log_tag) {
+      for (const i in styles) {
+        const formatter = formatterResolver(styles[i], color, date, log_tag);
+        const fmt = formatter.formatObject(data);
+        process.stdout.write(`${fmt}\n`);
       }
     }
+
+    // if (data.log_tag) {
```
Comment (alkiskal) [inline lib/appenders/console.js]: should these be removed?

File: lib/formatters/BaseFormatter.js [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    let prefix = `[${obj.requestId}] ${inReq.MSG} ${obj.method} ${obj.path}`;
+    if (this.appendDate) {
+      prefix = `[${new Date().toISOString()}] ${prefix}`;
+    }
+    if (this.color) {
+      prefix = chalk[[colors[inReq.TAG]]](prefix);
+    }
+    return prefix;
+
+    //return `${prefix}`;
```
Comment (alkiskal) [inline lib/formatters/BaseFormatter.js]: remove

Comment (manoskouvarakis) [reply lib/formatters/BaseFormatter.js]: Done

File: lib/appenders/console.js (hunk 14 above)
Comment (manoskouvarakis) [reply lib/appenders/console.js]: Yes! Done

File: lib/adapters/default.js (hunk 13 above)
Comment (manoskouvarakis) [reply lib/adapters/default.js]: Done

**Review by alkiskal — APPROVED** [review]

**Pull request merged**